lib/Mojolicious/Guides/Cookbook.pod

```diff
@@ -1413,12 +1413,15 @@ to process requests in smaller batches.
     # Stop if there are no more URLs
     return unless my $url = shift @urls;
 
+    # Save __SUB__ for later use in another callback
+    my $fetch = __SUB__;
```

Comment:
How is this cleaner than `my $fetch; $fetch = sub {`?

Reply:
ok, i am lost.

I simple wanted to fix a bug in the documentation. My initial fix was not accepted, because of a newly introduced memleak.

The optimized non-memleak variant has now the problem it looks to similar to the memleak-variant?

Can someone please help me and tell me what exactly is required to fix this bug? I wanted a minimalistic change not a complete rewrite of the example.

Next try would be to rename the inner $fetch to something else, can someone please suggest a name, i am tired of trying to help ...

Reply:
Then just don't use `__SUB__` in an example where it serves no purpose.

Reply:
If you don't want to figure out the correct solution, please open an issue instead of a PR in the future.

Reply:
> Then just don't use `__SUB__` in an example where it serves no purpose.

@jhthorsen can you please confirm the difference between `my $fetch ; $fetch = sub { ...` and `my $fetch = sub { ... my $fetch = __SUB__; ...`

In my understanding the later (current MR) will not leak a reference after the `$promise->wait;`

Reply:
> If you don't want to figure out the correct solution, please open an issue instead of a PR in the future.

Ofc i want a correct solution. I spend time to write this MR, rewrite it and ask questions to get this done.



But i really would appreciate some hints when someone pointed to a problem.

> Inconsistent comment format.

I am not a native speaker, so i lost about 15 minutes to figure out what do you mean (and i am still not sure if i fixed this correctly).

And with your `How is this cleaner than my $fetch; $fetch = sub {?` i am again thinking that i misread/understood also this question. Can you give a hint why you asked this after @jhthorsen told that this is wrong?

Reply:
@Sadrak 

> can you please confirm the difference between `my $fetch ; $fetch = sub { ...` and `my $fetch = sub { ... my $fetch = __SUB__; ...`

If I understand correctly, they are looking for:

```perl
my $fetch; $fetch = sub {

  # Stop if there are no more URLs
  return unless my $url = shift @urls;

  # Fetch the next title
  $ua->get($url => sub ($ua, $tx) {
    print "$url: ", $tx->result->dom->at('title')->text, "\n";

    # Next request
    $fetch->();
    $promise->resolve if --$count == 0;
  });
  $count++;
};
```

`$fetch` must be declared first, and then assigned to, so that it is available within the function.  If it was assigned during the declaration `my $fetch = sub { ...` then `$fetch` would not be available within the body of the function.

Reply:
pretty sure this solution get a downvote from @jhthorsen (that was my first try).

> Pretty sure that introduces a circular reference, so -1 from me.

And he is absolut right.

So to resume: The documentation is broken. I tried to figure this out, but can't get a proper/accepted solution. Next step is what? I guess closing and ignoring? :-/ That hurt in my developer heart ...

Reply:
I lost track of the history of this discussion and have no idea what @jhthorsen was referring to.

Reply:
using `my $fetch ; $fetch = sub { ...; $fetch->(); ...; }` will leak one reference. `$fetch` won't cleaned up after the scope due to a circular reference.

Alternatively before calling `$promise->resolve` we have to call `undef $fetch` manually to break the circular reference.

Or use the current MR with `my $fetch = __SUB__` as a solution to this problem.

Reply:
In my testing so far, `my $fetch; $fetch =` is flagged as circular by `Devel::Cycle`.  But then so is $ua.   Declaring `my $fetch = __SUB__` within the sub does remove that circular reference, but `$ua` remains unless it is passed in as an argument, e.g.

```perl
  my $fetch = sub ($ua) {
    return unless my $url = shift @urls;

    my $fetch = __SUB__;
    printf "refcount: %s\n", refcount($fetch);

    $ua->get(
      $url => sub ($ua, $tx) {
        print "$url: ", $tx->result->json->{result}, "\n";

        $fetch->($ua);
        $promise->resolve if --$count == 0;
      }
    );

    $count++;
  };

  $fetch->($ua) for 1 .. 5;
  $promise->wait;
```

I'm still testing for alternatives, and also to understand why the above example finishes with a high refcount, whereas the version with `my $fetch; $fetch =` finishes with a refcount of 1.